src/main/java/eu/europeana/metis/sandbox/service/metrics/MetricsServiceImpl.java

```diff
@@ -0,0 +1,226 @@
+    progressDatasetEntity.setTotalRecords(report.getTotalRecords());
+    progressDatasetEntity.setStatus(report.getStatus().value());
+    progressDatasetEntity.setEndTimeStamp(LocalDateTime.now());
+    progressDatasetRepository.save(progressDatasetEntity);
+  }
+
+  private ProgressDatasetEntity findProgressDataset(String datasetId) {
+    return progressDatasetRepository.findByDatasetId(datasetId);
+  }
+
+  private ProgressStepEntity findProgressStep(String datasetId, Step step) {
+    return progressStepRepository.findByDatasetIdAndStep(datasetId, step.value());
+  }
+
+  private void getProgressDatasetMetrics(Map<String, Object> metricsMap) {
+    List<ProgressDatasetEntity> progressDatasetEntities = progressDatasetRepository.findAll();
+
+    LongSummaryStatistics totalStatistics =
+        progressDatasetEntities.stream()
+                               .mapToLong(ProgressDatasetEntity::getTotalRecords)
+                               .summaryStatistics();
+
+    LongSummaryStatistics processedStatistics =
+        progressDatasetEntities.stream()
+                               .mapToLong(ProgressDatasetEntity::getProcessedRecords)
+                               .summaryStatistics();
+
+    LongSummaryStatistics durationStatistics =
+        progressDatasetEntities.stream()
+                               .mapToLong(
+                                   item -> Duration.between(item.getStartTimeStamp(),
+                                       item.getEndTimeStamp()).toSeconds())
+                               .summaryStatistics();
+    Map<String, Object> master = new HashMap<>();
+    addMetricToOverview(totalStatistics, master, "TotalRecords");
+    addMetricToOverview(processedStatistics, master, "ProcessedRecords");
+    addMetricToOverview(durationStatistics, master, "Duration");
+    master.put("DatasetCount", progressDatasetRepository.count());
+    metricsMap.put("MetricsByDataset", master);
+  }
+
+  private void getProgressStepMetrics(Map<String, Object> metricsMap) {
+    List<ProgressStepEntity> progressStepEntities = progressStepRepository.findAll();
+    Map<String, Object> stepCategory = new HashMap<>();
+    stepMetrics(stepCategory, progressStepEntities, Step.HARVEST_ZIP);
+    stepMetrics(stepCategory, progressStepEntities, Step.HARVEST_OAI_PMH);
+    stepMetrics(stepCategory, progressStepEntities, Step.VALIDATE_EXTERNAL);
+    stepMetrics(stepCategory, progressStepEntities, Step.TRANSFORM);
+    stepMetrics(stepCategory, progressStepEntities, Step.VALIDATE_INTERNAL);
+    stepMetrics(stepCategory, progressStepEntities, Step.NORMALIZE);
+    stepMetrics(stepCategory, progressStepEntities, Step.ENRICH);
+    stepMetrics(stepCategory, progressStepEntities, Step.MEDIA_PROCESS);
+    stepMetrics(stepCategory, progressStepEntities, Step.PUBLISH);
+    metricsMap.put("MetricsByStep", stepCategory);
+  }
+
+  private void stepMetrics(Map<String, Object> category, List<ProgressStepEntity> progressStepEntities, Step step) {
```

Comment:
Some goes here about the modification/extension of the parameter.

Reply:
also here refactored.